Add `get_area_def` to cf reader

BENR0 opened the pull request:
<!-- Describe what your PR does, and why -->
This adds add area definition support for the cf reader. I fixed the tests and added a test area in geos projection as well as some preliminary assertions for testing the area.
I think the tests should be improved and extended to cover other areas. Currently the extent assertion is failing because of https://github.com/pytroll/pyresample/issues/355.

<!-- For works in progress choose "Create draft pull request" from the drop-down green button. -->

 - [x] Closes #1672 <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->

djhoese commented:
Pyresample 1.20. 0 is being deployed right now. It looks like our test environment uses pyresample from conda-forge so we'll have to wait for that package to be pull requested, merged, released, and synced to the package repository.

mraspaud commented:
This is failing, so it'll have to go to the next release

djhoese commented:
@zxdawn @BENR0 @mraspaud So I just ran this locally with pyresample main and it does actually fail even though the CI jobs here aren't pulling in the right version:

```
>           assert expected_area.area_extent == actual_area.area_extent
E           AssertionError: assert (339045.5577,... 4803645.4685) == (4803645.4685..., 339045.5577)
E             At index 0 diff: 339045.5577 != 4803645.468500001
E             Use -v to get the full diff

satpy/tests/reader_tests/test_satpy_cf_nc.py:144: AssertionError
```

BENR0 commented:
It seems this was due to a switch of x and y in the test setup. Now it passes. One caveat though is that I had to add `pytest.approx` to the assert statement because the upper right y coordinate was just a tiny bit off from the expected value.

djhoese commented:
You may need to merge `main` into this branch. It seems github actions doesn't like the old workflow file (I hope that's the problem at least).

BENR0 commented:
Indeed I totally forgot to do that.

BENR0 commented:
@djhoese @mraspaud did you have time to look at this?

mraspaud commented:
@BENR0 sorry for the late reply.

It looks good for the area case, but what happens if we are working with swath definitions (so we just have lons and lats)? will the reading crash then?

BENR0 commented:
I think that depends if this is covered by the implementations in pyresample. I can have a look at it even though I am not that familiar with the code. Maybe @TomLav can help because most of the code is from him.

mraspaud commented:
@BENR0 I just meant to verify that the changes you make do not change the behaviour of the lon/lat case, no need to implement it here. So if you can see that the lon/lat case still works as before, then this is good as it is now

BENR0 commented:
@mraspaud do I understand correctly that you mean if a netcdf written from a Satpy Scene which only has a swath definition is also read correctly?

mraspaud commented:
I haven't checked recently, but I would expect so.

mraspaud commented:
Although it might be the same problem as you are fixing here that the geolocation isn't actually read.

TomLav commented:
Thanks for tagging me. I am maybe wondering if cf_reader should offer a richer interface to pyresample's from_cf() routine? For example, from_cf() offers the option to request the area_def attached to a specific netCDF variable which is not available from the cf_reader() at present.
To specify variable= in from_cf() can useful both if there are several area_defs in a netCDF file (not illegal, but not often) and if the CF encoding is not fully standard (in which case specifying variable= helps from_cf() to locate the area_def).
But I do not know if the cf_reader workflow only looks at file-level area_defs or variable-level.

TomLav commented:
Thanks. Your approach is fine with me.

BENR0 commented:
@mraspaud regarding the swath definition case: You were right implementing the `get_area_def` the way I did breaks this case. The reason is that the CF utils in pyresample implemented by @TomLav do not cover that case. Without the `get_area_def` method implemented in the specific reader Satpy "falls back" to creating the area from the lat/lon grids which is implemented in the `FileYAMLReader`. For a quick work around I added `ValueError` to the try except in order to fall back to the yaml reader creating the area def. While this works I think this is basically asking for trouble since `ValueError` is pretty generic and might lead to more problems that it solves.

My opinion about this is that handling 2D lat/lon coordinate grids should also be implemented in the CF utils in pyresample because CF allows this kind of coordinates and therefore `AreaDefinition.from_cf` should be able to handle this.
Apart from this obviously there is a unit test missing in the cf reader for the 2D lat/lon grid case. Are there some helper functions somewhere already to create a Satpy dataset with a swath definition which could be reused?

TomLav commented:
Hello. I think it is correct of pyresample's `from_cf()` to not support lat/lon swath data (if this is what was failing now). I think AreaDefinition is only for grids based on Earth projections, while the geometry information of lat/lon swath would use SwathDefinition object.

I don't know satpy. Can you link me to the current implementation of this feature in satpy, so that I can see how it is working?

BENR0 commented:
You are correct that 2D lat/lon  grids are handled by `SwathDefinition`. The formulation with `AreaDefinition` I used was misleading. What I was trying to say is that it would be good to have a `from_cf` method that can handle both (so returns either `AreaDefinition` or `SwathDefinition` objects depending on the case).

The logic for the creation of the area the `satpy_cf_nc` reader used before the changes proposed in this PR is located in the `FileYAMLReader`. I think https://github.com/pytroll/satpy/blob/cd057caee2b48f718eff2bc3b384e6286ced420b/satpy/readers/yaml_reader.py#L831 might be a starting point if you want to have a look at it. The try/except handles the fall back to creating a `SwathDefinition` if the `get_area_def` method is not implemented in the reader.

TomLav commented:
I see. I'll give it some thoughts. A first step would be to have a from_cf() method for pyresample's Geometry, that would default returning a SwathDefinition if no valid CF grid construct was detected.

djhoese commented:
@BENR0 @TomLav Any ideas for my previous questions:

> I'm hesitant to want to add SwathDefinition support directly to from_cf in pyresample. Does the CF standard define standard ways of defining arrays of lon/lat coordinates? What do we do if there are multiple? This seems like something will be different for every file scheme (thinking of all of the readers in Satpy that have different names for their lon/lat variables). ...actually CF doesn't standardize variable names so it would have to be done by standard_name at least.

This is still my opinion, that pyresample's `from_cf` should not support generating SwathDefinitions unless there is a clear definition of how lon/lats are to be defined and referenced from CF standard NetCDF variables.

Unless I'm reading the code incorrectly, the previous behavior always tried to generate a SwathDefinition from the "coordinates" defined in the YAML. This is the same as any polar-orbiter reader in Satpy. I think you should try/except in `get_area_def` and raise a `NotImplementedError` instead of the `ValueError` that is generated now (and remove the new ValueError check in the YAML reader). I know `NotImplementedError` may seem like the wrong thing to do here, but historically this is what we've done and catching ValueError could cause other issues in the future. For example, an actual ValueError trying to generate an AreaDefinition that should have been possible to generate.

BENR0 commented:
@djhoese I changed the try/except.

I am good with not implementing something regarding SwathDefinitions in `from_cf`. The way it is now netcdf files written by Satpy can be read either way which is all I need for now.
I agree that the more general approach has to many open questions/is to difficult and can/should be something for another PR if it is desired in the future.

djhoese commented:
pre-commit.ci autofix

djhoese commented:
This looks pretty good to me. I'm not sure how I feel about the warning as in some cases that may be the expected and intended path forward for the user. For example, if they configured coordinates in the YAML then they probably know that an AreaDefinition can't be made from the NetCDF file. @mraspaud what do you think?

BENR0 commented:
@djhoese @mraspaud what's the status on merging this?

mraspaud commented on satpy/readers/satpy_cf_nc.py:
```diff
@@ -294,8 +295,21 @@ def get_dataset(self, ds_id, ds_info):
+            # No CF compliant projection information was found in the netcdf file or
+            # file contains 2D lat/lon arrays. To fall back to generating a SwathDefinition
+            # with the yaml_reader NotImplementedError is raised.
+            logger.warning("Failed to load AreaDefinition from nc file. Falling back to SwathDefinition.")
```
I think this should be an info rather than a warning, and maybe reading "No AreaDefinition to load from nc file. Falling back to SwathDefinition."

BENR0 replied:
Ok I will change it.

BENR0 replied:
Just saw your comment above: yes it is falling back to SwathDefinition. I am embarrassed to say I can't remember if I added a unit test for this on another computer. I will check on that. But it is definitely tested manually ;-).

djhoese replied:
I kind of feel this should be a DEBUG level. In the cases that the file/reader is customized with the coordinates that make up the longitude/latitude then there is no reason for any "extra" log messages. In these cases, this is expected behavior.

mraspaud replied:
Makes sense. @BENR0 sorry for asking for yet another change!

BENR0 replied:
No problem. It's an easy change ;-).

mraspaud reviewed (commented):
This look good, just one comment if I may. Also just to make sure, is the falling back to SwathDefintion tested?

mraspaud reviewed (approved):
LGTM

mraspaud commented:
@BENR0 could you check the SwathDef test?

BENR0 commented:
@mraspaud I added the test for the SwathDefinition. It seems to me that the this test module could use some refactoring because there is some repetitive code in setting up datasets for different tests. To keep it inline with the overall style I also added duplicate code for generating a dataset. I am not sure but maybe this could all be moved into the setup (setting up a Scene with a dataset for each of SwathDefinition, AreaDefinition and for prefix testing and then in the test just load the needed dataset?).

Also I think with #2176 the instrument attribute setting can be removed for the dataset writing.

mraspaud commented:
sounds good, feel free to refactor!

BENR0 commented:
@mraspaud I refactored the dataset generation to be in the set up for all tests. I left in the writing of the header attribute `instrument` because #2176 is not merged yet and thus tests would fail.

mraspaud commented:
#2176 is now merged @BENR0

BENR0 commented:
@mraspaud ok will update this pr later today or tomorrow morning.

BENR0 commented:
@mraspaud I merged main and refactored out the instrument lines in the tests.

mraspaud commented:
@BENR0 regarding the failing tests on windows, I can recommend the usage of the `tmp_path` fixture, maybe that can help.

djhoese commented:
@BENR0 Any time to look into the failing tests?

BENR0 commented:
@djhoese yes I looked at it and it seems that the failing test on windows is due to the same filename being used in two tests. I don't know if @mraspaud hint with the `tmp_path` fixture can fix that. But it seems that in later tests file pathes are just counted up  to make them different. I can just use that same technique here if that is ok with you.

mraspaud commented:
Yes the fixture seems to work better on windows than the tempdir functions

djhoese commented:
@BENR0 if a single test function is writing multiple files then counting and produce unique filenames seems reasonable. Although this is probably a sign that the function should be split into multiple tests.

Using `tmp_path` as the destination for the files you create in the tests should mean that each test gets its own unique directory and so one test shouldn't interfere with another.

BENR0 commented:
@djhoese @mraspaud since the tests are using unittest.testcase the pytest `tmp_path` fixture does not work. I actually think it would be nice to use it but that would need some refactoring. Haven't used the `tmp_path` fixture before so could be I am missing something and there is a workaround somehow.

Maybe I will have some time next week to work on this. If this should move forward faster I could just do the hacky filename numbering in the mean time.

djhoese commented:
Give me 5 seconds. I'll fix it.

djhoese commented:
@BENR0 I've made a PR to your fork that should do the pytest and tmp_path changes. It cleans up the code a ton:

https://github.com/BENR0/satpy/pull/1

djhoese commented:
Note I assume that if the filenames are in a separate tmp directory for each test that the filenames can be the same for each test. If that means the tests are no longer testing what they were before then we can make multiple fixtures for each filename format.

BENR0 commented:
:-D after I wrote my comment I started the same changes you made and tested it for the first test. But didn't have the time to finish it. Thanks for the work.

djhoese commented:
Yeah, no problem. I'm avoiding my other work so this was a nice distraction and I've pytest'd quite a few test modules at this point. All you have to do now is merge the PR in your fork and the tests here *should* pass...I hope.

mraspaud commented:
@djhoese @BENR0 nice with the switch to pytest!

djhoese commented:
The unstable test failure seems to be a problem in rasterio...I hope. I'm going to try restarting the test and see what happens. If it fails again I think I'll merge this anyway and then work on figuring out what broke in a separate PR.

djhoese commented:
Filed https://github.com/rasterio/rasterio/issues/2591 with rasterio. Let's merge this and worry about the failing test somewhere else.
